sycl/doc/SpecializationConstants.md

```diff
@@ -0,0 +1,305 @@
+`program::set_spec_constant<IDType>(val)` calls from the app.
+
+**_NOTE_**  `IDType` gets translated to the symbolic ID using the integration
+header mechanism, similarly to kernel ID type. The reason why
+`__builtin_unique_stable_name` is not used here is because this code is
+compiled by the host compiler, which can be any C++ 14-compatible compiler
+unaware of the clang-specific built-ins.
+
+Before JIT-ing a program, the runtime "flushes" the spec constants: it iterates
+through the value map and invokes the
+
+```cpp
+pi_result piextProgramSetSpecializationConstant(pi_program prog,
+                                                pi_uint32 spec_id,
+                                                size_t spec_size,
+                                                const void *spec_value);
+```
+
+Plugin Interface function for each entry, taking the `spec_id` from the ID map.
+
+## "Plain Old Data" (POD) types support design
+
+#### Source representation
+
+Say, the POD type is
+
+```cpp
+struct A {
+  int x;
+  float y;
+};
+
+struct POD {
+  A a[2];
+  int b;
+};
+```
+
+and the user says
+
+```cpp
+  POD gold{
+    {
+      { goldi, goldf },
+      { goldi + 1, goldf + 1 },
+    },
+    goldi
+  };
+
+  cl::sycl::ONEAPI::experimental::spec_constant<POD, MyConst> sc =  program4.set_spec_constant<MyConst>(gold);
+```
+
+#### Compiler
+
+##### The SpecConstant pass changes
+
+ - The SpecConstants pass in the post-link will have the following IR as input (`sret` conversion is omitted for clarity):
+
+```cpp
+  %spec_const = call %struct.POD __sycl_getCompositeSpecConstantValue<mangling for POD type template specialization ("MyConst_mangled")
+```
+
+where `__sycl_getCompositeSpecConstantValue` is a new "intrinsic"
+ (in addition to `__sycl_getSpecConstantValue`) recognized by SpecConstants pass,
+ which creates a value of a composite (of non-primitive type) specialization constant.
+ It does not need a default value, because its default value consists of default
+ valued of its leaf specialization constants (see below).
+
+ - after spec constant enumeration (symbolic -\> int ID translation), the SC pass
+ will handle the `__sycl_getCompositeSpecConstantValue`. Knowning the composite
+ specialization constant's type (`%struct.POD`), the pass will traverse its leaf
+ fields and generate 5 "primitive" spec constants using already existing SPIR-V intrinsic:
+
+```cpp
+%gold_POD_a0x = call i32 __spirv_SpecConstant(i32 10, i32 0)
+%gold_POD_a0y = call float __spirv_SpecConstant(i32 11, float 0)
+%gold_POD_a1x = call i32 __spirv_SpecConstant(i32 12, i32 0)
+%gold_POD_a1y = call float __spirv_SpecConstant(i32 13, float 0)
+%gold_POD_b = call i32 __spirv_SpecConstant(i32 14, i32 0)
+```
+
+And 1 "composite"
+
+```cpp
+  %gold_POD = call %struct.POD __spirvCompositeSpecConstant<POD mangling>(10, 11, 12, 13, 14, 15)
+```
+
+where `__spirvCompositeSpecConstant<type mangling>` is a new SPIR-V intrinsic which
```

Comment:
There are at least 3 different ways to denote mangling of the POD type <in angle brackets> in this document. I believe they all mean the same thing, can we have single representation for it?

Reply:
OK